### ISSUE TITLE
Question: velocity js

### ISSUE THREAD

**Issue by dcworldwide** [timbrandin/blaze-react#41]
I have been using velocity js within my on rendered code. Do you think it would work as is with your solution?

**Comment by timbrandin** [comment]
I'm not sure, what are you testing?

**Comment by faceyspacey** [comment]
@dcworldwide Velocity the popular animation framework http://julian.com/research/velocity ..It's become the replacement for jQuery's animation utilities lately. 

...basically, there are various React addons for animation--though the main one listed on their site is pretty weak. I'll eventually be releasing the equivalent of `Meteor.Components` which will let you choose either Blaze as the renderer or React. Blaze and react share a similar API that notifies you when dom elements are added to the screen, eg:in Blaze: `_uihooks.insertElement`. (it's an *unofficial* API in Blaze by the way.) And in my Meteor React Component's package you in fact can specify animations for when elements are added/removed via velocity style sequences (http://julian.com/research/velocity/#uiPack) for advanced animations.

Doing complete free-style animation is another story, but I suspect it will be solved for React as well if it hasn't already, which means we can write an adapter. Writing such adapters is way easier than writing this whole transpiler. So basically if we can achieve the foundation, you can rest assured we or someone will do the same for things like velocity. I think at the end of the day if you animate elements and put them write where React expects them (just like Blaze has you do with their unofficial animation api) it should work. And if that doesn't work out-the-box, I'm sure that's a problem someone in the react ecosystem is perfecting.

**Comment by patrickml** [comment]
@faceyspacey React already has animations built in using https://facebook.github.io/react/docs/animation.html instead of using JS for the animations it using css by adding and removing classes over time allowing hardware accleration to be using on animations.

**Comment by faceyspacey** [comment]
That's exactly the built-in add on I was mentioning. And it's extremely weak. You won't find yourself chaining sequences of animations dynamically there.

**Comment by patrickml** [comment]
But couldn't you used the javascript api for react animation to work like blazes uihooks?

From Docs: 

ReactTransitionGroup is the basis for animations. It is accessible from require('react-addons-transition-group'). When children are declaratively added or removed from it (as in the example above) special lifecycle hooks are called on them.

**componentWillAppear(callback)**
This is called at the same time as componentDidMount() for components that are initially mounted in a TransitionGroup. It will block other animations from occurring until callback is called. It is only called on the initial render of a TransitionGroup.

**componentDidAppear()**
This is called after the callback function that was passed to componentWillAppear is called.

**componentWillEnter(callback)**
This is called at the same time as componentDidMount() for components added to an existing TransitionGroup. It will block other animations from occurring until callback is called. It will not be called on the initial render of a TransitionGroup.

**componentDidEnter()**
This is called after the callback function that was passed to componentWillEnter is called.

**componentWillLeave(callback)**
This is called when the child has been removed from the ReactTransitionGroup. Though the child has been removed, ReactTransitionGroup will keep it in the DOM until callback is called.

**componentDidLeave()**
This is called when the willLeave callback is called (at the same time as componentWillUnmount).

In other words you could do 
```jsx
Template.name.animation({
   componentWillAppear : function () {
      $(this.refs.someRef).velocity("transition.fadeIn")
   }
})
```

I don't know its just an idea i havent tested it out Im just throwing it out there

**Comment by faceyspacey** [comment]
there are some issues to make it work properly, but there are packages working on this problem:

https://github.com/whattokingu/react-animation-mixin
https://github.com/chenglou/react-motion
https://github.com/Nutelac/react.animate

The first 2 seem to work with dynamic animation properties, and the last one adds full on support for dynamic scene building by animating state properties instead of pure style properties. I wonder how React will hold up performance wise when all the child components have to be re-rendered as well (at least to virtual DOM). All the child components' render methods will all be called. If you have a big tree of components, that's a lot of code being executed. Luckily it's not code affecting the actual DOM. If you're using the pure render mixin for all the child components, perhaps the only `render` method being re-run is that of the component for which you're animating its state. enjoy!

**Comment by dcworldwide** [comment]
Thanks for the discussion, this is a big react-migration pain point for my velocity.js heavy app. 

Just to clarify our use case, as @faceyspacey suggests, velocity.js's value is in its ability to compose complex sequences of animation with fine grained control through a well defined js interface. For instance, a velocity js flow could scroll to a grid item, then shake the item, then expand it into a modal, then render the modal template, by progressively staggering in the modal's contents. It's pretty neat in terms of the UX it can achieve but most importantly, the code to achieve such logic is a single block of pure javascript and is simple to maintain compared to a css solution.
